narwhal/consensus/src/bullshark.rs

```diff
@@ -314,4 +329,40 @@ impl Bullshark {
         // Return its certificate and the certificate's digest.
         dag.get(&round).and_then(|x| x.get(&leader))
     }
+
+    /// Updates and calculates the reputation score for the current commit managing any internal state.
+    /// It returns the updated reputation score.
+    fn update_reputation_score(
+        &mut self,
+        state: &mut ConsensusState,
+        committed_sequence: &[Certificate],
+        next_sub_dag_index: u64,
+    ) -> ReputationScores {
+        // we reset the scores for every schedule change window.
+        // TODO: when schedule change is implemented we should probably change a little bit
+        // this logic here.
+        if next_sub_dag_index % self.change_schedule_every_committed_sub_dags == 0 {
+            state.last_consensus_reputation_score.clear();
```

Comment:
for now, if we want to use the score for the transaction submission, it may be more instructive to Sui to use a rolling window, rather than a clear

Reply:
That will probably lead us to a different implementation compared to what I've done here and will require to switch to a buffering approach. As per our previous discussions having this as a window that changes every K-commits guarantees more predictable/deterministic behaviour across the nodes, as they'll have the opportunity to use the same scores for a longer period (thus making things more predictable across nodes)

Reply:
Ok, if this is the case then, we may want to start with some non-zero score as a default so that when we use the score in other places, we don't need to write special cases for when we receive a newly-reset score.

Reply:
Ok let's get back to this idea, but before this could you please have a look on the [final_of_schedule](https://github.com/MystenLabs/sui/pull/8581/files#diff-974cd380877235846592d528fbfbd4addcfedfc83ec546e8f12548c98076e2c1R77) property I've introduced on the `ReputationScores` struct? That will help us easily/agnostically handle the score updates on the consumer side.

Reply:
Makes sense to have a more coarse grained metric than a rolling window because if we change scores every round the client behavior during submission might look unpredictable from outside. However we might want to initialise scores at the beginning of the epoch as well for validators that are known to be slow

Reply:
> However we might want to initialise scores at the beginning of the epoch as well for validators that are known to be slow
Since we reset the scores per new schedule, why do we want to do that? I mean on the schedule we'll basically restart the score counting process as a way to evaluate the data of this window.

Reply:
to avoid paying the cost of equal reputation during the first k DAGs although we know who is probably still slow

Reply:
Let's continue this discussion offline and I can address changes in follow up PR